host/linux/posix.c

```diff
@@ -0,0 +1,1133 @@
+
+    return 0;
+}
+
+/*
+**==============================================================================
+**
+** Polling:
+**
+**==============================================================================
+*/
+
+typedef struct _wait_args
+{
+    int64_t enclaveid;
+    oe_host_fd_t epfd;
+    int maxevents;
+    struct epoll_event events[];
+} wait_args_t;
+
+static void* _epoll_wait_thread(void* arg_)
+{
+    int ret;
+    wait_args_t* args = (wait_args_t*)arg_;
+    int retval;
+
+    ret = epoll_wait((int)args->epfd, args->events, args->maxevents, -1);
+
+    if (ret >= 0)
+    {
+        size_t num_notifications = (size_t)ret;
+        struct epoll_event* ev = args->events;
+        oe_device_notifications_t* notifications =
+            (oe_device_notifications_t*)ev;
+
+        OE_STATIC_ASSERT(sizeof(notifications[0]) == sizeof(ev[0]));
+
+        if (oe_posix_polling_notify_ecall(
+                (oe_enclave_t*)args->enclaveid,
+                &retval,
+                notifications,
+                num_notifications) != OE_OK)
+        {
+            goto done;
+        }
+
+        if (retval != 0)
+            goto done;
+    }
+
+done:
+    free(args);
+    return NULL;
+}
+
+typedef struct _poll_args
+{
+    int64_t enclaveid;
+    oe_host_fd_t epfd;
+    nfds_t nfds;
+    struct pollfd fds[];
+} poll_args_t;
+
+static void* _poll_wait_thread(void* arg_)
+{
+    int ret;
+    poll_args_t* args = (poll_args_t*)arg_;
+    int retval;
+
+    ret = poll(args->fds, args->nfds, -1);
+    if (ret >= 0)
+    {
+        size_t num_notifications = (size_t)ret;
+        struct pollfd* ev = args->fds;
+        oe_device_notifications_t* notifications =
+            (oe_device_notifications_t*)ev;
+
+        size_t ev_idx = 0;
+        size_t notify_idx = 0;
+        for (ev_idx = 0; ev_idx < (size_t)args->nfds; ev_idx++)
+        {
+            if (ev[ev_idx].revents)
+            {
+                notifications[notify_idx].events = (uint32_t)ev[ev_idx].revents;
+                notifications[notify_idx].data.list_idx = (uint32_t)ev_idx;
+
+                /* ATTN: casting 64-bit fd to 32-bit fd. */
+                notifications[notify_idx].data.epoll_fd = (int)args->epfd;
+            }
+        }
+
+        if (oe_posix_polling_notify_ecall(
+                (oe_enclave_t*)args->enclaveid,
+                &retval,
+                notifications,
+                num_notifications) != OE_OK)
+        {
+            goto done;
+        }
+
+        if (retval != 0)
+            goto done;
+    }
+
+done:
+    free(args);
+    return NULL;
+}
+
+oe_host_fd_t oe_posix_epoll_create1_ocall(int flags)
+{
+    errno = 0;
+
+    return epoll_create1(flags);
+}
+
+int oe_posix_epoll_wait_async_ocall(
+    int64_t enclaveid,
+    oe_host_fd_t epfd,
+    size_t maxevents)
+{
+    int ret = -1;
+    size_t eventsize;
+    pthread_t thread = 0;
+    wait_args_t* args = NULL;
+
+    eventsize = sizeof(struct oe_epoll_event) * maxevents;
```

Comment:
@EMumau Seems like epoll implementation here could benefit from switchless calling.

Reply:
Yes, that would be nice. At the very least we should cache some host memory.